File: src/nu/tunedal/bdmem/BDMemWidget.java
package nu.tunedal.bdmem;

import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.ContentResolver;
import android.content.Context;
import android.database.Cursor;
import android.provider.ContactsContract;
import android.util.Pair;
import android.widget.RemoteViews;
import android.widget.TextView;
import android.widget.LinearLayout;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import android.text.TextUtils;
import android.provider.ContactsContract.CommonDataKinds.Event;

public class BDMemWidget extends AppWidgetProvider {
    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager,
                         int[] appWidgetIds) {
        System.out.println("Sweet zombie Jesus!");
        for (int i=0; i < appWidgetIds.length; i++) {
            RemoteViews views = new RemoteViews(context.getPackageName(),
                                                R.layout.bdmem_appwidget);
            views.removeAllViews(R.id.container);
            for (Pair<String,String> p : getBirthdays(context)) {
                RemoteViews row = new RemoteViews(context.getPackageName(),
                                                  R.layout.widget_row);
                row.setTextViewText(R.id.datum, p.first);
                row.setTextViewText(R.id.namn, p.second);
                views.addView(R.id.container, row);
                System.out.println("getBirthdays: " + p.first
                                   + ", " + p.second);
            }
            // TODO: Uppdatera med Service istället, eftersom det kan
            // ta en stund att köra queryn.
            /*
            System.out.println("Sleeping...");
            try {
                Thread.sleep(30000);
            }
            catch (InterruptedException ex) {
            }
            System.out.println("Awake!");
            */
            appWidgetManager.updateAppWidget(appWidgetIds[i], views);
        }
    }

    public List<Pair<String,String>> getBirthdays(Context context) {
        String[] projection = {
            ContactsContract.Data._ID,
            Event.CONTACT_ID,
            Event.START_DATE,
            Event.TYPE,
            ContactsContract.Data.MIMETYPE
        };

        ContentResolver resolver = context.getContentResolver();
        String where = (ContactsContract.Data.MIMETYPE +
                        " = ? AND " +
                        ContactsContract.Data.DATA2 +
                        " = ?");
        Cursor cur = resolver.query(ContactsContract.Data.CONTENT_URI,
                                    projection,
                                    where,
                                    new String[] {
                                        Event.CONTENT_ITEM_TYPE,
                                        "" + Event.TYPE_BIRTHDAY
                                    },
                                    ContactsContract.Data._ID + " ASC");
        List<Pair<String,String>> list = new ArrayList<Pair<String,String>>();
        list.add(Pair.create("Rader:", "" + cur.getCount()));
        int j = 1;
        String[] ignorables = {"name", "phone_v2", "email_v2", "photo"};
        Map<String,String> birthdays = new HashMap<String,String>();
        Map<String,String> nicknames = new HashMap<String,String>();
        while (cur.moveToNext() && j++ < 5) {
            birthdays.put(cur.getString(1), cur.getString(2));
            boolean ignore = false;
            for (String s : ignorables) {
                if (cur.getString(projection.length - 1).endsWith("/" + s)) {
                    ignore = true;
                    break;
                }
            }
            if (ignore) continue;
            StringBuilder sb = new StringBuilder();
            for (int i=0; i < projection.length; i++) {
                sb.append(", ");
                sb.append(cur.getString(i));
            }
            System.out.println(sb.toString());
        }
        System.out.println(TextUtils.join(", ", birthdays.keySet()));
        // TODO: query mot kontakttabellen för att översätta id:n till namn.
        for (Map.Entry<String,String> entry : birthdays.entrySet()) {
            list.add(Pair.create(entry.getKey(), entry.getValue()));
        }
        return list;
    }
}

File: app/src/main/java/nu/tunedal/bdmem/BDMemWidget.kt
package nu.tunedal.bdmem

import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.Context
import android.database.Cursor
import android.provider.ContactsContract
import android.widget.RemoteViews
import android.provider.ContactsContract.CommonDataKinds.Event
import java.util.Calendar
import android.app.PendingIntent
import android.content.Intent
import android.graphics.Color
import android.util.Log

class CursorIterator(val cursor: Cursor): Iterator<Cursor> {
    private var valid = cursor.moveToFirst().also { cursor.move(-1) }
    override fun next() = cursor.apply { valid = moveToNext() }
    override fun hasNext() = valid && !cursor.isLast
}

operator fun Cursor.iterator() = CursorIterator(this)

fun <T: Any> Iterator<T>.consume(count: Int): Sequence<T> {
    var i = 0
    return generateSequence { if (i++ < count && hasNext()) next() else null }
}

data class Birthday(val name: String, val date: String) {
    val birthday get() = date.substring(5)
}

fun <T> MutableList<T>.rotate(offset: Int) {
    java.util.Collections.rotate(this, offset)
}

fun <T> Collection<T>.rotated(offset: Int): List<T> =
        toMutableList().apply { rotate(offset) }

val Calendar.month get() = get(Calendar.MONTH) + 1
val Calendar.dayOfMonth get() = get(Calendar.DAY_OF_MONTH)

object log {
    private const val tag = "BDMem"
    fun verbose(msg: String) = Log.v(tag, msg)
    fun debug(msg: String) = Log.d(tag, msg)
    fun info(msg: String) = Log.i(tag, msg)
    fun warning(msg: String) = Log.w(tag, msg)
    fun error(msg: String) = Log.e(tag, msg)
}

class BDMemWidget : AppWidgetProvider() {
    override fun onUpdate(context: Context,
                          appWidgetManager: AppWidgetManager,
                          appWidgetIds: IntArray) {
        log.debug("Sweet zombie Jesus!")
        for (widgetId in appWidgetIds) {
            val views = RemoteViews(context.packageName,
                    R.layout.bdmem_appwidget)
            views.removeAllViews(R.id.container)

            val updateIntent = Intent().apply {
                action = AppWidgetManager.ACTION_APPWIDGET_UPDATE
                putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS,
                         intArrayOf(widgetId))
            }
            val pendingIntent = PendingIntent.getBroadcast(
                    context, 0, updateIntent,
                    PendingIntent.FLAG_UPDATE_CURRENT)
            views.setOnClickPendingIntent(R.id.container, pendingIntent)

            val showPast = 1
            val lines = 5
            val birthdays = getBirthdays(context, showPast).iterator().let {
                val past = it.consume(showPast).map { it to false }
                val future = it.consume(lines - showPast).map { it to true }
                past + future
            }

            for ((birthday, upcoming) in birthdays) {
                val (namn, datum) = birthday
                val row = RemoteViews(context.packageName,
                        R.layout.widget_row)
                if (!upcoming) {
                    val color = Color.GRAY
                    row.setTextColor(R.id.datum, color)
                    row.setTextColor(R.id.namn, color)
                }
                row.setTextViewText(R.id.datum, datum)
                row.setTextViewText(R.id.namn, namn)
                views.addView(R.id.container, row)
                log.verbose("getBirthdays: ${datum}, ${namn}")
            }
            // TODO: Uppdatera med Service istället.
            appWidgetManager.updateAppWidget(widgetId, views)
        }
    }

    fun getBirthdays(context: Context, past: Int = 0): List<Birthday> {
        val now = Calendar.getInstance().let {
            "%02d-%02d".format(it.month, it.dayOfMonth)
        }
        val projection = arrayOf(
                ContactsContract.Data._ID,
                Event.CONTACT_ID,
                ContactsContract.Contacts.DISPLAY_NAME,
                Event.START_DATE,
                Event.TYPE,
                ContactsContract.Data.MIMETYPE)
        val resolver = context.contentResolver
        val where = ContactsContract.Data.MIMETYPE +
                " = ? AND " +
                ContactsContract.Data.DATA2 +
                " = ?"
        val cursor = resolver.query(ContactsContract.Data.CONTENT_URI,
                projection,
                where,
                arrayOf(Event.CONTENT_ITEM_TYPE, "" + Event.TYPE_BIRTHDAY),
                null)
        val birthdays = cursor.use { c ->
            c.iterator().asSequence().map {
                Birthday(it.getString(2), it.getString(3))
            }.sortedBy { it.birthday }.toList()
        }
        if (birthdays.isEmpty())
            return emptyList()
        val nextIndex = birthdays.indexOfFirst { it.birthday >= now }
                .let { if (it == -1) 0 else it }  // första om ingen hittas
        log.debug("Next birthday: index $nextIndex, ${birthdays[nextIndex]}")
        return birthdays.rotated(past - nextIndex)
    }
}
